test/chart-template.js

```diff
@@ -0,0 +1,392 @@
+/* global Highcharts */
+(function (global) {
+
+    /* *
+     *
+     *  Private Properties
+     *
+     * */
+
+    /**
+     * The chart template registry
+     *
+     * @type {Dictionary<ChartTemplate>}
+     */
+    var chartTemplates = {};
+
+    /* *
+     *
+     *  Private Functions
+     *
+     * */
+
+    /**
+     * Creates a new container in the DOM tree.
+     *
+     * @return {HTMLElement}
+     * The DOM element of the container
+     */
+    function createContainer() {
+        var container = global.document.createElement('div');
+        container.style.left = '0';
+        container.style.positon = 'absolute';
+        container.style.top = '0';
+        global.document.body.appendChild(container);
+        return container;
+    }
+
+    /**
+     * Creates a deep copy of entries and properties.
+     *
+     * @param {array|object} source
+     * The source to copy from.
+     *
+     * @param {object} propertiesTree
+     * The properties tree to copy.
+     *
+     * @return {array|object}
+     * The copy of the source.
+     */
+    function treeCopy(source, propertiesTree) {
+        if (!source) {
+            return source;
+        }
+        switch (typeof source) {
+            default:
+                return source;
+            case 'array':
+                return JSON.parse(JSON.stringify(source));
```

Comment:
Won't that be a problem with dataset like this? 

```javascript
{
  series: [{
    data: [{
      y: 10,
      events: { click: function () { } }
    }]
  }]
}
```

`stringify()` will remove callback. Similar for array of series, where a series has events.

Reply:
This would be only a problem, if you like to create templates with events. In the test you can create event functions and they will get removed in the end of the test case.

Reply:
Sorry, let me explain, for example I want to test multiple yAxes with events:

```javascript
    ChartTemplate.test('line-simple', {
        yAxis: [{
            events: {
                afterSetExtremes: function () {
                    this.chart.firedCallback1 = true;
                }
            }
        }, {
            events: {
                afterSetExtremes: function () {
                    this.chart.firedCallback2 = true;
                }
            }
        }],
        series: [{
            data: [1, 2]
        }]

    }, function (template) {

        var chart = template.chart;

        chart.yAxis[0].setExtremes(0, 10000, false);
        chart.yAxis[1].setExtremes(0, 10000);

        assert.strictEqual(
            chart.firedCallback1 && chart.firedCallback2,
            true,
            'Events are fine'
        );
    });
```

Demo is fine: http://jsfiddle.net/BlackLabel/o5san9w6/3/

Test failed:
> HeadlessChrome 0.0.0 (Mac OS X 10.13.3) Highcharts series-line/type FAILED
>	Events are fine
>	Expected: true
>	Actual: undefined

Reply:
If I apply the same update functionality of the template to your fiddle, it fails too: http://jsfiddle.net/o5san9w6/7/
I am not sure, why it breaks - especially with the first y-axis. Chart.update should support axis updates and fire events.

Reply:
Ha! Looks like I found a bug in attaching/removing events for the yAxis..

Reply:
I have integrated the bugfix #8161. Please retry your tests.